### PR TITLE
Rewrite stable diffusion unet with com.microsoft attenion operators

### PR DESCRIPTION
**Pull request by titaiwangms**
This PR introduces attention rewriting rules for stable diffusion unet. In summary, scripted function _scaled_dot_product_attention is replaced by either cim.microsoft.Attention, or com.microsoft.MultiHeadAttention.

### CODE REVIEW

File: onnxscript/rewriter/onnxruntime/transformers/multihead_attention.py [hunk 1]
```diff
@@ -101,9 +133,9 @@ def __init__(self) -> None:
 
     @_version_controller.register_version(min_version="4.33", max_version="4.36")
     def _fusion_with_4d_cache(self, function: ir.Function) -> ir.Function:
-        if len(function.input) != 9:
+        if len(function.inputs) != 9:
```
Comment (justinchuby) [inline onnxscript/rewriter/onnxruntime/transformers/multihead_attention.py:136]: Add a comment on what 9 is?

File: onnxscript/rewriter/onnxruntime/transformers/multihead_attention.py [hunk 2]
```diff
@@ -592,3 +623,89 @@ def phi_attention(
...
+    FUNCTION_KEYWORD = "Attention"
+    PACKAGE_NAME = "diffusers"
+    _version_controller = function_rule.VersionController()
+
+    def __init__(self) -> None:
+        super().__init__()
+
+    @_version_controller.register_version()
+    def _fusion(self, function: ir.Function) -> ir.Function:
+        if len(function.inputs) != 6 and len(function.inputs) != 7:
```
Comment (justinchuby) [inline onnxscript/rewriter/onnxruntime/transformers/multihead_attention.py:638]: Explain where this 6/7 inputs is defined? Perhaps a link to transformers code etc?

File: onnxscript/rewriter/onnxruntime/transformers/multihead_attention.py [hunk 3]
```diff
@@ -55,39 +55,71 @@
...
+                    head_size = constant_numpy_value[3]
+                    return AttnSizeConfig(
+                        num_attention_heads=num_attention_heads,
+                        num_key_value_heads=None,
+                        head_size=head_size,
+                        hidden_size=hidden_size,
+                    )
             raise function_rule.FunctionRewriteError(
-                f"Unexpected number of outputs. Expected 3, got {len(function.outputs)}."
+                "Failed to infer head size and number of heads from Reshape nodes."
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply onnxscript/rewriter/onnxruntime/transformers/multihead_attention.py]: Done

File: onnxscript/rewriter/onnxruntime/transformers/multihead_attention.py [hunk 4]
```diff
@@ -55,39 +55,71 @@
...
-        num_key_value_heads = present_value.shape[1]
-        hidden_size = attn_output.shape[2]
-        num_attention_heads = hidden_size // head_size
-        return AttnSizeConfig(
-            num_attention_heads=num_attention_heads,
-            num_key_value_heads=num_key_value_heads,
-            head_size=head_size,
-            hidden_size=hidden_size,
+        raise function_rule.FunctionRewriteError(
+            f"Unexpected function structure, got output: {len(function.outputs)}."
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply onnxscript/rewriter/onnxruntime/transformers/multihead_attention.py]: Done

File: onnxscript/rewriter/onnxruntime/transformers/multihead_attention.py (hunk 1 above)
Comment (titaiwangms) [reply onnxscript/rewriter/onnxruntime/transformers/multihead_attention.py:142]: I will leave this to @BowenBao

File: onnxscript/rewriter/onnxruntime/transformers/multihead_attention.py (hunk 2 above)
Comment (titaiwangms) [reply onnxscript/rewriter/onnxruntime/transformers/multihead_attention.py:648]: I added the expected input names.

File: onnxscript/rewriter/onnxruntime/transformers/multihead_attention.py [hunk 5]
```diff
@@ -55,39 +55,71 @@
...
-    num_key_value_heads: int
+    num_key_value_heads: int | None
     head_size: int
     hidden_size: int
 
 
 class AttentionRewriteRule(function_rule.FunctionRewriteRule, abc.ABC):
     def infer_attn_size_config(self, function: ir.Function) -> AttnSizeConfig:
-        if len(function.outputs) != 3:
+        if len(function.outputs) == 3:
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply onnxscript/rewriter/onnxruntime/transformers/multihead_attention.py:73]: Added description

**Review by justinchuby — APPROVED** [review]

File: onnxscript/rewriter/onnxruntime/transformers/multihead_attention.py [hunk 6]
```diff
@@ -55,39 +55,77 @@
...
-        num_key_value_heads = present_value.shape[1]
-        hidden_size = attn_output.shape[2]
-        num_attention_heads = hidden_size // head_size
-        return AttnSizeConfig(
-            num_attention_heads=num_attention_heads,
-            num_key_value_heads=num_key_value_heads,
-            head_size=head_size,
-            hidden_size=hidden_size,
+        raise function_rule.FunctionRewriteError(
+            f"Attenion modules should have 3 outputs or scaled_dot_product_attention node, \
```
Comment (justinchuby) [inline onnxscript/rewriter/onnxruntime/transformers/multihead_attention.py:127]:
```suggestion
            f"Attenion modules should have 3 outputs or scaled_dot_product_attention node,
```

File: onnxscript/rewriter/onnxruntime/transformers/multihead_attention.py [hunk 7]
```diff
@@ -592,3 +629,93 @@ def phi_attention(
...
+
+class MHAStableDiffusionUnetRewriteRule(AttentionRewriteRule):
+    """Rewrite rule for Attention in diffusers."""
+
+    FUNCTION_KEYWORD = "Attention"
+    PACKAGE_NAME = "diffusers"
+    _version_controller = function_rule.VersionController()
+
+    def __init__(self) -> None:
+        super().__init__()
```
Comment (justinchuby) [inline onnxscript/rewriter/onnxruntime/transformers/multihead_attention.py:642]:
```suggestion
```

**Pull request merged**